File: Blind 75/13 - 1D Dynamic Programming/70-climbing-stairs.py
'''
You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
'''
from time import time


class Solution:
    def __init__(self):
        self.dic = {1: 1, 2: 2}

    def climbStairs(self, n: int) -> int:
        if n not in self.dic:
            self.dic[n] = self.climbStairs(n - 1) + self.climbStairs(n - 2)
        return self.dic[n]

    def reference(self, n: int) -> int:
        # Pointers will always be initialized to 1 and 1 during bottom-up
        # as there is only 1 way to land at n when starting at n (0 step) or n - 1 (1 step)
        p1, p2 = 1, 1

        for _ in range(n - 1):
            temp = p1
            p1 += p2
            p2 = temp

        return p1

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.climbStairs(case))
                else:
                    self.climbStairs(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [2, 3, 4]
    test.quantify(test_cases)

File: Blind 75/13 - 1D Dynamic Programming/JavaScript/70-climbing-stairs.js
const { performance } = require('perf_hooks');

class Solution {
  method(n) {
    const steps = { 1: 1, 2: 2 };
    const dfs = (n) => {
      if (!(n in steps)) steps[n] = dfs(n - 1) + dfs(n - 2);
      return steps[n];
    };

    return dfs(n);
  }

  reference(n, index = 0) {
    const isBaseCase1 = n < index;
    if (isBaseCase1) return 0;

    const isBaseCase2 = index === n;
    if (isBaseCase2) return 1;

    const [next, nextNext] = [index + 1, index + 2];
    const left = this.reference(n, next); /* Time O(2^N) | Space O(N) */
    const right = this.reference(n, nextNext); /* Time O(2^N) | Space O(N) */

    return left + right;
  }

  dfs = (char, graph, seen, buffer) => {
    if (seen.has(char)) return seen.get(char);

    if (!this.backTrack(char, graph, seen, buffer)) return false;

    buffer.push(char);

    return true;
  };

  backTrack = (char, graph, seen, buffer) => {
    seen.set(char, false);
    for (const neighbor of graph.get(char)) {
      if (!this.dfs(neighbor, graph, seen, buffer)) return false;
    }
    seen.set(char, true);

    return true;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [2, 3];
test.quantify(testCases);
